pkg/destroy/volume.go

```diff
@@ -0,0 +1,266 @@
+	logger                    logrus.FieldLogger
+
+	mu sync.Mutex
+}
+
+func newKubeClientSet(kubeConfig string) (*kubernetes.Clientset, error) {
+	auth, err := getKubeConfigAuth(kubeConfig)
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not get auth kubeconfig")
+	}
+	config, err := clientcmd.RESTConfigFromKubeConfig(auth)
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not get config from kubeconfig")
+	}
+	return kubernetes.NewForConfig(config)
+}
+
+// NewVolume sets up the Volume struct, first configuring the kubernetes client set, then
+// retrieving the persistent volumes and their claims.
+func NewVolume(ctx context.Context, logger logrus.FieldLogger, kubeConfig string) (*Volume, error) {
+	v := &Volume{}
+	ctx, cancel := context.WithTimeout(ctx, 60*time.Second)
+	defer cancel()
+
+	clientSet, err := newKubeClientSet(kubeConfig)
+	if err != nil {
+		return nil, err
+	}
+	v.clientSet = clientSet
+
+	pvcList, err := clientSet.CoreV1().PersistentVolumeClaims("").List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+	v.persistentVolumeClaimList = pvcList.DeepCopy()
+
+	pvList, err := clientSet.CoreV1().PersistentVolumes().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+	v.persistentVolumeList = pvList.DeepCopy()
+	v.logger = logger
+
+	return v, nil
+}
+
+func (v *Volume) drainNodes(ctx context.Context) error {
+	timeout, cancel := context.WithTimeout(ctx, 15*time.Minute)
+	defer cancel()
+
+	v.logger.Info("Draining and cordoning nodes, this might take a few moments.")
+	drainHelper := &drain.Helper{
+		Ctx:                             timeout,
+		Client:                          v.clientSet,
+		Force:                           true,
+		GracePeriodSeconds:              0,
+		IgnoreAllDaemonSets:             true,
+		Timeout:                         0,
+		DeleteEmptyDirData:              true,
+		Selector:                        "",
+		PodSelector:                     "",
+		ChunkSize:                       0,
+		DisableEviction:                 true,
+		SkipWaitForDeleteTimeoutSeconds: 0,
+		AdditionalFilters:               nil,
+		// do we need any of this output?
+		Out:                             io.Discard,
+		ErrOut:                          io.Discard,
+		DryRunStrategy:                  0,
+		OnPodDeletedOrEvicted:           nil,
+		OnPodDeletionOrEvictionFinished: func(pod *corev1.Pod, usingEviction bool, err error) {},
+		OnPodDeletionOrEvictionStarted:  func(pod *corev1.Pod, usingEviction bool) {},
+	}
+
+	nodeList, err := v.clientSet.CoreV1().Nodes().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	// Need all nodes cordoned before draining so there is no
+	// available resources for the pod to restart
+	for _, node := range nodeList.Items {
+		if _, ok := node.Labels["node-role.kubernetes.io/worker"]; ok {
+			v.logger.Debugf("cordoning node: %s", node.Name)
+			if err := drain.RunCordonOrUncordon(drainHelper, &node, true); err != nil {
```

Comment:
Can we really drain/cordon all the nodes at the same time? Some nodes with PDB will refuse to be drained, I guess. Since we are doing this right at the beginning of the destroy operation, will this make cluster "unstable" (we cordoned all the nodes!).

When I was testing similar thing for statefulsets, where there is only one node in a zone(topology), the drain operation causes a lot of `Pending` pods, because replacement pods have nowhere to be scheduled.  Usually pvc protection controller permits deletion of PVCs used by pending pods, but the interaction is weird. :(